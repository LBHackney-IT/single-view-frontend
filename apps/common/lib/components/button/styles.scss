@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-button/button";

.button-loading-indicator {
  display: inline-block;
  border-top: 2px solid currentcolor;
  border-right: 2px solid currentcolor;
  border-bottom-style: solid;
  border-left-style: solid;
  border-radius: 50%;
  border-bottom-width: 2px;
  border-left-width: 2px;
  border-bottom-color: transparent;
  border-left-color: transparent;
  width: 1em;
  height: 1em;
  color: currentColor;
  animation: 0.45s linear 0s infinite normal none running
    button-loading-indicator-rotation;
  margin-right: 1em;
  vertical-align: middle;

  span {
    border: 0px;
    clip: rect(0px, 0px, 0px, 0px);
    height: 1px;
    width: 1px;
    margin: -1px;
    padding: 0px;
    overflow: hidden;
    white-space: nowrap;
    position: absolute;
  }
}

.lbh-button--chevron.govuk-button {
  padding: 15px 24px;
  @include lbh-rem(font-size, 27);

  & > svg {
    margin-top: 0;
    margin-left: 6px;
  }
}

@keyframes button-loading-indicator-rotation {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}
