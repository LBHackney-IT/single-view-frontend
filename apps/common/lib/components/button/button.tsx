import React, { forwardRef } from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

const AddIcon = () => {
  return (
    <svg width="12" height="12" viewBox="0 0 12 12">
      <path d="M6.94 0L5 0V12H6.94V0Z" />
      <path d="M12 5H0V7H12V5Z" />
    </svg>
  );
};

const ChevronIcon = () => {
  return (
    <svg
      width="20"
      height="22"
      viewBox="0 0 20 22"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M0 0H8.54573L20 10H11.4543L0 0Z"
        fill="white"
      />
      <path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M0 22L8.54573 22L20 10H11.4543L0 22Z"
        fill="#96CCAE"
      />
    </svg>
  );
};

export interface ButtonProps {
  variant?: "primary" | "secondary" | "chevron" | "add";
  isLoading?: boolean;
  isDisabled?: boolean;
  loadingText?: string;
  override?: number;
}

export type ButtonComponent = Polymorphic.ForwardRefComponent<"button", ButtonProps>;

export const Button: ButtonComponent = forwardRef(function Button(
  {
    as: ButtonComp = "button",
    variant = "primary",
    isLoading = false,
    loadingText,
    isDisabled,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const buttonClasses = cn(
    "govuk-button",
    "lbh-button",
    {
      "govuk-button--primary lbh-button--add": variant === "add",
      "lbh-button--chevron": variant === "chevron",
      "govuk-button--secondary lbh-button--secondary": variant === "secondary",
      "lbh-button--disabled govuk-button--disabled": isDisabled,
    },
    widthOverrides(override),
    className,
  );

  const disabled = isDisabled || isLoading || undefined;

  return (
    <ButtonComp
      ref={ref}
      className={buttonClasses}
      type={ButtonComp === "button" ? "button" : undefined}
      disabled={ButtonComp === "button" ? disabled : undefined}
      aria-disabled={disabled}
      {...props}
    >
      {variant === "add" && !isLoading && <AddIcon />}
      {isLoading && (
        <span className="button-loading-indicator">
          <span>Loading...</span>
        </span>
      )}
      {isLoading && loadingText ? loadingText : children}
      {variant === "chevron" && <ChevronIcon />}
    </ButtonComp>
  );
});
