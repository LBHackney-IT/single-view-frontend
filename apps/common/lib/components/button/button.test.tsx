import React, { useEffect } from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";

import { Button } from "./button";

test("it renders correctly", async () => {
  const { container } = render(<Button>Test Button</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it shows a loading state", async () => {
  const { container } = render(<Button isLoading>Test Button</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it shows a loading state with custom text", () => {
  render(
    <Button isLoading loadingText="Processing">
      Test Button
    </Button>,
  );
  expect(screen.getByText("Processing")).toBeInTheDocument();
});

test("it renders the correct variant", async () => {
  const { container } = render(<Button variant="secondary">Test Link</Button>);
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it can render as a link", async () => {
  const { container } = render(
    <Button as="a" href="/">
      Test Link
    </Button>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it can render an 'add' variant", async () => {
  const { container } = render(<Button variant="add">Add icon variant</Button>);
  expect(container).toMatchSnapshot();
});

test("it can render an 'add' variant (without icon if loading)", async () => {
  const { container } = render(
    <Button isLoading variant="add">
      Add icon variant
    </Button>,
  );
  expect(container).toMatchSnapshot();
});

test("it can render a 'chevron' variant", async () => {
  const { container } = render(<Button variant="chevron">Chevron icon variant</Button>);
  expect(container).toMatchSnapshot();
});

test("it accepts a ref", () => {
  const callback = jest.fn();
  const Comp = () => {
    const ref = React.createRef<HTMLButtonElement>();

    useEffect(() => {
      if (ref.current) {
        callback(ref.current);
      }
    }, [ref]);
    return <Button ref={ref}>Test Button</Button>;
  };
  render(<Comp />);
  expect(callback).toBeCalledTimes(1);
});
