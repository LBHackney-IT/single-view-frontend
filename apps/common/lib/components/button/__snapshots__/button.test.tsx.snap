// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it can render a 'chevron' variant 1`] = `
<div>
  <button
    class="govuk-button lbh-button lbh-button--chevron"
    type="button"
  >
    Chevron icon variant
    <svg
      fill="none"
      height="22"
      viewBox="0 0 20 22"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M0 0H8.54573L20 10H11.4543L0 0Z"
        fill="white"
        fill-rule="evenodd"
      />
      <path
        clip-rule="evenodd"
        d="M0 22L8.54573 22L20 10H11.4543L0 22Z"
        fill="#96CCAE"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</div>
`;

exports[`it can render an 'add' variant (without icon if loading) 1`] = `
<div>
  <button
    aria-disabled="true"
    class="govuk-button lbh-button govuk-button--primary lbh-button--add"
    disabled=""
    type="button"
  >
    <span
      class="button-loading-indicator"
    >
      <span>
        Loading...
      </span>
    </span>
    Add icon variant
  </button>
</div>
`;

exports[`it can render an 'add' variant 1`] = `
<div>
  <button
    class="govuk-button lbh-button govuk-button--primary lbh-button--add"
    type="button"
  >
    <svg
      height="12"
      viewBox="0 0 12 12"
      width="12"
    >
      <path
        d="M6.94 0L5 0V12H6.94V0Z"
      />
      <path
        d="M12 5H0V7H12V5Z"
      />
    </svg>
    Add icon variant
  </button>
</div>
`;

exports[`it can render as a link 1`] = `
<div>
  <a
    class="govuk-button lbh-button"
    href="/"
  >
    Test Link
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <button
    class="govuk-button lbh-button"
    type="button"
  >
    Test Button
  </button>
</div>
`;

exports[`it renders the correct variant 1`] = `
<div>
  <button
    class="govuk-button lbh-button govuk-button--secondary lbh-button--secondary"
    type="button"
  >
    Test Link
  </button>
</div>
`;

exports[`it shows a loading state 1`] = `
<div>
  <button
    aria-disabled="true"
    class="govuk-button lbh-button"
    disabled=""
    type="button"
  >
    <span
      class="button-loading-indicator"
    >
      <span>
        Loading...
      </span>
    </span>
    Test Button
  </button>
</div>
`;
